backend/dataall/modules/dataset_sharing/services/share_managers/s3_bucket_share_manager.py

```diff
@@ -0,0 +1,443 @@
+            kms_target_resources = [
+                f"arn:aws:kms:{self.bucket_region}:{self.source_account_id}:key/{kms_key_id}",
+                f"arn:aws:kms:{self.bucket_region}:{self.source_account_id}:key/{kms_key_id}/*"
+            ]
+
+            share_manager.add_missing_resources_to_policy_statement(
+                resource_type=kms_key_id,
+                target_resources=kms_target_resources,
+                existing_policy_statement=existing_policy["Statement"][1],
+                iam_role_policy_name=IAM_S3BUCKET_ROLE_POLICY
+            )
+
+            policy = existing_policy
+        else:
+            logger.info(
+                f'{IAM_S3BUCKET_ROLE_POLICY} does not exists for IAM role {self.target_requester_IAMRoleName}, creating...'
+            )
+            policy = {
+                "Version": "2012-10-17",
+                "Statement": [
+                    {
+                        "Effect": "Allow",
+                        "Action": [
+                            "s3:*"
+                        ],
+                        "Resource": [
+                            f"arn:aws:s3:::{self.bucket_name}",
+                            f"arn:aws:s3:::{self.bucket_name}/*"
+                        ]
+                    },
+                    {
+                        "Effect": "Allow",
+                        "Action": [
+                            "kms:*"
+                        ],
+                        "Resource": [
+                            f"arn:aws:kms:{self.bucket_region}:{self.source_account_id}:key/{kms_key_id}",
+                            f"arn:aws:kms:{self.bucket_region}:{self.source_account_id}:key/{kms_key_id}/*"
+                        ]
+                    }
+                ]
+            }
+        IAM.update_role_policy(
+            self.target_account_id,
+            self.target_requester_IAMRoleName,
+            IAM_S3BUCKET_ROLE_POLICY,
+            json.dumps(policy),
+        )
+
+    def get_bucket_policy_or_default(self):
+        """
+        Fetches the existing bucket policy for the S3 bucket if one exists otherwise returns the default bucket policy
+        :return:
+        """
+        s3_client = S3Client(self.source_account_id, self.source_environment.region)
+        bucket_policy = s3_client.get_bucket_policy(self.bucket_name)
+        if bucket_policy:
+            logger.info(
+                f'There is already an existing policy for bucket {self.bucket_name}, will be updating policy...'
+            )
+            bucket_policy = json.loads(bucket_policy)
+        else:
+            logger.info(
+                f'Bucket policy for {self.bucket_name} does not exist, generating default policy...'
+            )
+            exceptions_roleId = self.get_bucket_owner_roleid()
+            bucket_policy = S3ControlClient.generate_default_bucket_policy(self.bucket_name, exceptions_roleId)
+        return bucket_policy
+
+    def get_bucket_owner_roleid(self):
```

Comment:
maybe move this function to share_manager_utils? could also be re-used in `s3_access_point_share_manager` in multiple places